pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java

```diff
@@ -17,239 +17,230 @@
+ *     <p>Once we have the inverted index buffers, we simply go over them and create the bitmap for each dictId and
+ *     serialize them into a file.
  *   </li>
  * </ul>
+ * <p>Based on the number of values we need to store, we use direct memory or MMap file to allocate the buffer.
  */
 public final class OffHeapBitmapInvertedIndexCreator implements InvertedIndexCreator {
-  // Maximum number of values due to 2GB file size limit
-  public static final int MAX_NUM_VALUES = Integer.MAX_VALUE / V1Constants.Numbers.INTEGER_SIZE;
+  // Use MMapBuffer if the buffer size is larger than 100MB
+  private static final int NUM_VALUES_THRESHOLD_FOR_MMAP_BUFFER = 25_000_000;
+
+  private static final String FORWARD_INDEX_VALUE_BUFFER_SUFFIX = ".fwd.idx.val.buf";
+  private static final String FORWARD_INDEX_LENGTH_BUFFER_SUFFIX = ".fwd.idx.len.buf";
+  private static final String INVERTED_INDEX_VALUE_BUFFER_SUFFIX = ".inv.idx.val.buf";
+  private static final String INVERTED_INDEX_LENGTH_BUFFER_SUFFIX = ".inv.idx.len.buf";
 
-  private final String _columnName;
+  private final File _invertedIndexFile;
+  private final File _forwardIndexValueBufferFile;
+  private final File _forwardIndexLengthBufferFile;
+  private final File _invertedIndexValueBufferFile;
+  private final File _invertedIndexLengthBufferFile;
   private final boolean _singleValue;
   private final int _cardinality;
   private final int _numDocs;
   private final int _numValues;
-  private final File _indexDir;
-  private final File _invertedIndexFile;
+  private final boolean _useMMapBuffer;
 
-  private ByteBuffer _valueByteBuffer;
-  private IntBuffer _valueIntBuffer;
+  // Forward index buffers (from docId to dictId)
+  private int _nextDocId;
+  private LBufferAPI _forwardIndexValueBuffer;
+  // For multi-valued column only because each docId can have multiple dictIds
+  private int _nextValueId;
+  private LBufferAPI _forwardIndexLengthBuffer;
 
-  // For multi-value column only
-  private int _valueBufferOffset;
-  private ByteBuffer _numValuesByteBuffer;
-  private IntBuffer _numValuesIntBuffer;
-
-  // Posting list related buffers
-  // Constructed in first pass
-  private ByteBuffer _postingListLengthByteBuffer;
-  private IntBuffer _postingListLengthIntBuffer;
-  // Constructed in second pass
-  private ByteBuffer _postingListValueByteBuffer;
-  private ByteBuffer _postingListCurrentOffsetByteBuffer;
+  // Inverted index buffers (from dictId to docId)
+  private LBufferAPI _invertedIndexValueBuffer;
+  private LBufferAPI _invertedIndexLengthBuffer;
 
   public OffHeapBitmapInvertedIndexCreator(File indexDir, FieldSpec fieldSpec, int cardinality, int numDocs,
-      int numValues) {
-    _columnName = fieldSpec.getName();
+      int numValues) throws IOException {
+    String columnName = fieldSpec.getName();
+    _invertedIndexFile = new File(indexDir, columnName + V1Constants.Indexes.BITMAP_INVERTED_INDEX_FILE_EXTENSION);
```

Comment:
I think indexDir is unique, just confiirming that if multiple segments of the same table are being loaded we won't collide on mmap files

Reply:
**indexDir** is per segment based